lib/drivers/insuletDriver.js

```diff
@@ -747,28 +1149,207 @@ module.exports = function (config) {
           .with_scheduleName(
           data.basalPrograms.names[data.basalPrograms.enabled_idx].name)
           .with_rate(basal.detail.basal_rate_units_per_hour)
-          // temporarily, we're going to set all scheduled basals to 1 day
-          .with_duration(24 * 60 * sundial.MIN_TO_MSEC)
           .with_time(basal.timestamp)
           .with_deviceTime(basal.deviceTime)
           .with_timezoneOffset(basal.timezoneOffset)
-          .set('index', data.log_records[basalrecs[b]].log_index)
+          .set('index', basal.log_index);
+      }
+      else {
+        postbasal = cfg.builder.makeTempBasal()
+          .with_rate(basal.detail.basal_rate_units_per_hour)
+          .with_time(basal.timestamp)
+          .with_deviceTime(basal.deviceTime)
+          .with_timezoneOffset(basal.timezoneOffset)
+          .with_duration(basal.detail.duration_msec)
+          .set('index', basal.log_index);
+        if (basal.detail.temp_basal_percent != null) {
+          var suppressed = cfg.builder.makeScheduledBasal()
+            .with_rate(common.fixFloatingPoint(basal.detail.basal_rate_units_per_hour/basal.detail.temp_basal_percent, 2))
+            .with_time(basal.timestamp)
+            .with_deviceTime(basal.deviceTime)
+            .with_timezoneOffset(basal.timezoneOffset)
+            .with_duration(basal.detail.duration_msec);
+          postbasal.with_percent(basal.detail.temp_basal_percent)
+            .set('suppressed', suppressed);
+        }
+      }
+
+      if (dupcheck[postbasal.time]) {
+        debug('Duplicate basals at ' + postbasal.deviceTime + '; expect PDM date & time settings change near here.');
+        // we can't push the "dups" until we can feed them to jellyfish as a separate stream
+        // postrecords.push(postbasal);
+      } else {
+        dupcheck[postbasal.time] = postbasal;
+        postrecords.push(postbasal);
+      }
+    }
+    var lastBasal = postrecords[postrecords.length - 1];
+    annotate.annotateEvent(lastBasal, 'insulet/basal/fabricated-from-schedule');
+    return records.concat(postrecords);
+  };
+
+  // it turns out these records may not actually be implemented
+  // occlusions are only represented through pump advisory and hazard alarms
+  // TODO: maybe we should just delete this code?
```

Comment:
Yeah, good idea. Probably also time to add my few additional comments on the spec - will talk to @brandonarbiter about doing that.